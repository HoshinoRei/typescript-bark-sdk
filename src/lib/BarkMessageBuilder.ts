import isUrl from "is-url"

import BarkMessageErrorType from "../model/enumeration/BarkMessageErrorType"
import type BarkMessageLevel from "../model/enumeration/BarkMessageLevel"
import type BarkMessageSound from "../model/enumeration/BarkMessageSound"
import BarkMessageError from "../model/error/BarkMessageError"
import type BarkMessage from "../model/request/BarkMessage"

/**
 * A class to help build Bark message properly
 * @property barkMessage The Bark message which will be built
 */
export default class BarkMessageBuilder {
  protected barkMessage: BarkMessage

  constructor() {
    this.barkMessage = {}
  }

  /**
   * Set the message as archived
   * @returns The builder
   */
  archive(): BarkMessageBuilder {
    this.barkMessage.isArchive = "1"
    return this
  }

  /**
   * Set the badge property of the message
   * @param badge
   * @returns The builder
   */
  badge(badge: number): BarkMessageBuilder {
    this.barkMessage.badge = badge
    return this
  }

  /**
   * When all the properties is set, use this function to build the message
   * @returns Bark message
   */
  build(): BarkMessage {
    return this.barkMessage
  }

  /**
   * Set the body property of the message
   * @param body
   * @returns The builder
   */
  body(body: string): BarkMessageBuilder {
    this.barkMessage.body = body
    return this
  }

  /**
   * Set the category of the message
   * @param category
   * @returns The builder
   */
  category(category: string): BarkMessageBuilder {
    this.barkMessage.category = category
    return this
  }

  /**
   * Set the copy property of the message
   * @param copy
   * @returns The builder
   */
  copy(copy: string): BarkMessageBuilder {
    this.barkMessage.copy = copy
    return this
  }

  /**
   * Set the device key property of the message
   * @param deviceKey
   * @returns The builder
   */
  deviceKey(deviceKey: string): BarkMessageBuilder {
    this.barkMessage.device_key = deviceKey
    return this
  }

  /**
   * Set the group property of the message
   * @param group
   * @returns The builder
   */
  group(group: string): BarkMessageBuilder {
    this.barkMessage.group = group
    return this
  }

  /**
   * Set the icon property of the message
   * @param icon
   * @returns The builder
   * @throws { @link BarkMessageError } if the argument is not a URL
   */
  icon(icon: string): BarkMessageBuilder {
    if (isUrl(icon)) {
      this.barkMessage.icon = icon
      return this
    } else {
      throw this.isNotUrlErrorProducer("icon")
    }
  }

  /**
   * Set the level property of the message
   * @param level
   * @returns The builder
   */
  level(level: BarkMessageLevel): BarkMessageBuilder {
    this.barkMessage.level = level
    return this
  }

  /**
   * Set the sound property of the message
   * @param sound
   * @returns The builder
   */
  sound(sound: BarkMessageSound): BarkMessageBuilder {
    this.barkMessage.sound = sound
    return this
  }

  /**
   * Set the title property of the message
   * @param title
   * @returns The builder
   */
  title(title: string): BarkMessageBuilder {
    this.barkMessage.title = title
    return this
  }

  /**
   * Set the URL property of the message
   * @param url
   * @returns The builder
   * @throws { @link BarkMessageError } if the argument is not a URL
   */
  url(url: string): BarkMessageBuilder {
    if (isUrl(url)) {
      this.barkMessage.url = url
      return this
    } else {
      throw this.isNotUrlErrorProducer("url")
    }
  }

  /**
   * An error producing function
   * @param argumentName The name of the argument which is not a URL
   * @returns Bark message error
   */
  protected isNotUrlErrorProducer(argumentName: string): BarkMessageError {
    return new BarkMessageError(
      BarkMessageErrorType.IS_NOT_URL,
      `The argument ${argumentName} is not a URL`,
    )
  }
}
