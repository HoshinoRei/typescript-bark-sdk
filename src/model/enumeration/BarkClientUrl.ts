enum BarkClientUrl {
  HEALTHZ = "/healthz",
  INFO = "/info",
  PUSH = "/push",
}

export default BarkClientUrl;
