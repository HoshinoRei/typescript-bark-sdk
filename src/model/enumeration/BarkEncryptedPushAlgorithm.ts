/**
 * The algorithms of Bark encrypted push
 */
enum BarkEncryptedPushAlgorithm {
  AES_128_CBC = "aes-128-cbc",
  AES_192_CBC = "aes-192-cbc",
  AES_256_CBC = "aes-256-cbc",
  AES_128_ECB = "aes-128-ecb",
  AES_192_ECB = "aes-192-ecb",
  AES_256_ECB = "aes-256-ecb",
}

export default BarkEncryptedPushAlgorithm;
