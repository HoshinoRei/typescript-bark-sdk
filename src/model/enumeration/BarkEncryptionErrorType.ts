/**
 * The types of Bark encryption error
 */
enum BarkEncryptionErrorType {
  KEY_IS_NOT_CORRECT,
  IV_IS_NOT_CORRECT,
}

export default BarkEncryptionErrorType
