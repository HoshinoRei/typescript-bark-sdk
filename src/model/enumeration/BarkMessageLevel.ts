/**
 * The levels of Bark message
 */
enum BarkMessageLevel {
  ACTIVE = "active",
  TIME_SENSITIVE = "timeSensitive",
  PASSIVE = "passive",
}

export default BarkMessageLevel
