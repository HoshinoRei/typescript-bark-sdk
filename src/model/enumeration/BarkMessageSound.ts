/**
 * The sounds of Bark message
 */
enum BarkMessageSound {
  ALARM = "alarm",
  ANTICIPATE = "anticipate",
  BELL = "bell",
  BIRDSONG = "birdsong",
  BLOOM = "bloom",
  CALYPSO = "calypso",
  CHIME = "chime",
  CHOO = "choo",
  DESCENT = "descent",
  ELECTRONIC = "electronic",
  FANFARE = "fanfare",
  GLASS = "glass",
  GOTOSLEEP = "gotosleep",
  HEALTHNOTIFICATION = "healthnotification",
  HORN = "horn",
  LADDER = "ladder",
  MAILSENT = "mailsent",
  MINUET = "minuet",
  MULTIWAYINVITATION = "multiwayinvitation",
  NEWMAIL = "newmail",
  NEWSFLASH = "newsflash",
  NOIR = "noir",
  PAYMENTSUCCESS = "paymentsuccess",
  SHAKE = "shake",
  SHERWOODFOREST = "sherwoodforest",
  SILENCE = "silence",
  SPELL = "spell",
  SUSPENSE = "suspense",
  TELEGRAPH = "telegraph",
  TIPTOES = "tiptoes",
  TYPEWRITERS = "typewriters",
  UPDATE = "update",
}

export default BarkMessageSound;
