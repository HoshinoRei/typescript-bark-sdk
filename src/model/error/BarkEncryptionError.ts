import type BarkEncryptionErrorType from "../enumeration/BarkEncryptionErrorType";

/**
 * Bark encryption error
 * @property type
 * @property message
 * @property cause
 */
export default class BarkEncryptionError extends Error {
  type: BarkEncryptionErrorType;
  message: string;
  cause?: Error;

  constructor(type: BarkEncryptionErrorType, message: string, cause?: Error) {
    super(message);
    this.type = type;
    this.message = message;
    this.cause = cause;
  }
}
