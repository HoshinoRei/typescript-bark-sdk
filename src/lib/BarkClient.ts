import axios, { AxiosError } from "axios";
import crypto from "crypto";

import BarkClientUrl from "../model/enumeration/BarkClientUrl";
import BarkEncryptedPushAlgorithm from "../model/enumeration/BarkEncryptedPushAlgorithm";
import BarkEncryptionErrorType from "../model/enumeration/BarkEncryptionErrorType";
import BarkResponseErrorType from "../model/enumeration/BarkResponseErrorType";
import BarkEncryptionError from "../model/error/BarkEncryptionError";
import BarkResponseError from "../model/error/BarkResponseError";
import type BarkMessage from "../model/request/BarkMessage";
import type BarkInfoResponse from "../model/response/BarkInfoResponse";
import type BarkResponse from "../model/response/BarkResponse";

/**
 * A class to communicate with Bark server
 */
export default class BarkClient {
  constructor(serverAddress: string = "https://api.day.app") {
    axios.defaults.baseURL = serverAddress;
  }

  /**
   * Check if the Bark server is healthy
   * @returns nothing if the Bark server is healthy
   * @see [Healthz](https://github.com/Finb/bark-server/blob/master/docs/API_V2.md#healthz)
   * @throws { @link BarkResponseError } if the Bark server is unhealthy
   */
  async health(): Promise<void> {
    try {
      await axios.get<string>(BarkClientUrl.HEALTHZ);
    } catch (e) {
      throw this.miscellaneousFunctionErrorProducer(e);
    }
  }

  /**
   * Get info of Bark server
   * @returns { @link BarkInfoResponse } info of the Bark server
   * @see [Info](https://github.com/Finb/bark-server/blob/master/docs/API_V2.md#info)
   * @throws { @link BarkResponseError } if the Bark server does not respond normally
   */
  async info(): Promise<BarkInfoResponse> {
    try {
      const { data } = await axios.get<{
        arch?: string;
        build?: string;
        commit?: string;
        devices?: number;
        version?: string;
      }>(BarkClientUrl.INFO);
      return {
        arch: data.arch,
        build:
          data.build !== undefined
            ? new Date(Date.parse(data.build))
            : undefined,
        commit: data.commit,
        devices: data.devices,
        version: data.version,
      };
    } catch (e) {
      throw this.miscellaneousFunctionErrorProducer(e);
    }
  }

  /**
   * Push a message to Bark APP
   * @param message bark message
   * @returns nothing if message is sent successfully
   * @see [API V2](https://github.com/Finb/bark-server/blob/master/docs/API_V2.md#api-v2)
   * @throws { @link BarkResponseError } if message is sent unsuccessfully
   */
  async push(message: BarkMessage): Promise<void> {
    try {
      await axios.post<BarkResponse>(BarkClientUrl.PUSH, message);
    } catch (e) {
      throw this.pushErrorProducer(e);
    }
  }

  /**
   * Push a message to Bark APP with encryption
   *
   * When using encrypted push, you do not need set device key in { @link BarkMessage }
   *
   * @param deviceKey device key
   * @param message bark message, whose device key is unneeded
   * @param algorithm which algorithm to use
   * @param key key
   * @param iv iv
   * @returns nothing if message is sent successfully
   * @see [Encryption](https://bark.day.app/#/en-us/encryption)
   * @throws { @link BarkResponseError } if message is sent unsuccessfully
   */
  async pushEncrypted(
    deviceKey: string,
    message: BarkMessage,
    algorithm: BarkEncryptedPushAlgorithm,
    key: string,
    iv: string,
  ): Promise<void> {
    this.checkKey(algorithm, key);
    this.checkIv(iv);

    const cipher = crypto.createCipheriv(algorithm, key, iv);

    try {
      await axios.post(
        deviceKey,
        {
          ciphertext: `${cipher.update(
            JSON.stringify(message),
            "utf-8",
            "base64",
          )}${cipher.final("base64")}`,
        },
        {
          headers: {
            "Content-Type": "application/x-www-form-urlencoded",
          },
        },
      );
    } catch (e) {
      throw this.pushErrorProducer(e);
    }
  }

  /**
   * An error producing function used by miscellaneous function
   * @param e error
   * @returns { @link BarkResponseError }
   */
  protected miscellaneousFunctionErrorProducer(e: unknown): BarkResponseError {
    if (e instanceof AxiosError) {
      return new BarkResponseError(
        BarkResponseErrorType.SERVER_HAS_NO_RESPONSE,
        "Server has no response",
        e,
      );
    } else if (e instanceof Error) {
      return new BarkResponseError(
        BarkResponseErrorType.UNKNOWN_ERROR,
        "Unknown error",
        e,
      );
    } else {
      return new BarkResponseError(
        BarkResponseErrorType.UNKNOWN_ERROR,
        "Unknown error",
      );
    }
  }

  /**
   * Check if key is legal when using encrypted push
   * @param algorithm which algorithm to use
   * @param key key
   * @throws { @link BarkEncryptionError } when key is illegal
   */
  protected checkKey(algorithm: BarkEncryptedPushAlgorithm, key: string): void {
    switch (algorithm) {
      case BarkEncryptedPushAlgorithm.AES_128_CBC:
      case BarkEncryptedPushAlgorithm.AES_128_ECB: {
        if (key.length !== 16) {
          throw new BarkEncryptionError(
            BarkEncryptionErrorType.KEY_IS_NOT_CORRECT,
            "The length of key is not 16",
          );
        }
        break;
      }
      case BarkEncryptedPushAlgorithm.AES_192_CBC:
      case BarkEncryptedPushAlgorithm.AES_192_ECB: {
        if (key.length !== 24) {
          throw new BarkEncryptionError(
            BarkEncryptionErrorType.KEY_IS_NOT_CORRECT,
            "The length of key is not 24",
          );
        }
        break;
      }
      case BarkEncryptedPushAlgorithm.AES_256_CBC:
      case BarkEncryptedPushAlgorithm.AES_256_ECB: {
        if (key.length !== 32) {
          throw new BarkEncryptionError(
            BarkEncryptionErrorType.KEY_IS_NOT_CORRECT,
            "The length of key is not 32",
          );
        }
        break;
      }
    }
  }

  /**
   * Check if iv is legal when using encrypted push
   * @param iv iv
   * @throws { @link BarkEncryptionError } when iv is illegal
   */
  protected checkIv(iv: string): void {
    if (iv.length !== 16) {
      throw new BarkEncryptionError(
        BarkEncryptionErrorType.IV_IS_NOT_CORRECT,
        "The length of iv is not 16",
      );
    }
  }

  /**
   * An error producing function used by push function
   * @param e error
   * @returns { @link BarkResponseError }
   */
  protected pushErrorProducer(e: unknown): BarkResponseError {
    if (e instanceof AxiosError) {
      const data: BarkResponse = e.response?.data;

      const failedToGetDeviceTokenRegularExpression =
        /failed to get device token: (.*)/;
      const requestBindFailedRegularExpression = /request bind failed: (.*)/;
      const pushFailedRegularExpression = /push failed: (.*)/;

      if (
        e.response?.status === 400 &&
        data.message === "device key is empty"
      ) {
        return new BarkResponseError(
          BarkResponseErrorType.DEVICE_KEY_IS_EMPTY,
          "Device key is empty",
          e,
        );
      } else if (
        e.response?.status === 400 &&
        data.message !== undefined &&
        failedToGetDeviceTokenRegularExpression.test(data.message)
      ) {
        const regularExpressionExecArray =
          failedToGetDeviceTokenRegularExpression.exec(data.message);
        return new BarkResponseError(
          BarkResponseErrorType.FAILED_TO_GET_DEVICE_TOKEN,
          regularExpressionExecArray != null
            ? `Failed to get device token: ${regularExpressionExecArray[1]}`
            : "Failed to get device token",
          e,
        );
      } else if (
        e.response?.status === 400 &&
        data.message !== undefined &&
        requestBindFailedRegularExpression.test(data.message)
      ) {
        const regularExpressionExecArray =
          requestBindFailedRegularExpression.exec(data.message);
        return new BarkResponseError(
          BarkResponseErrorType.REQUEST_BIND_FAILED,
          regularExpressionExecArray != null
            ? `Request bind failed: ${regularExpressionExecArray[1]}`
            : "Request bind failed",
          e,
        );
      } else if (
        e.response?.status === 500 &&
        data.message !== undefined &&
        pushFailedRegularExpression.test(data.message)
      ) {
        const regularExpressionExecArray = pushFailedRegularExpression.exec(
          data.message,
        );
        return new BarkResponseError(
          BarkResponseErrorType.PUSH_FAILED,
          regularExpressionExecArray != null
            ? `Push failed: ${regularExpressionExecArray[1]}`
            : "Push failed",
          e,
        );
      } else {
        return new BarkResponseError(
          BarkResponseErrorType.UNKNOWN_ERROR,
          "Unknown error",
          e,
        );
      }
    } else if (e instanceof Error) {
      return new BarkResponseError(
        BarkResponseErrorType.UNKNOWN_ERROR,
        "Unknown error",
        e,
      );
    } else {
      return new BarkResponseError(
        BarkResponseErrorType.UNKNOWN_ERROR,
        "Unknown error",
      );
    }
  }
}
