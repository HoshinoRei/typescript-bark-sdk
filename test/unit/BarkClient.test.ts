import { afterEach, describe, expect, test } from "@jest/globals";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import crypto from "crypto";

import BarkClient from "../../src/lib/BarkClient";
import BarkMessageBuilder from "../../src/lib/BarkMessageBuilder";
import BarkClientUrl from "../../src/model/enumeration/BarkClientUrl";
import BarkEncryptedPushAlgorithm from "../../src/model/enumeration/BarkEncryptedPushAlgorithm";
import BarkEncryptionErrorType from "../../src/model/enumeration/BarkEncryptionErrorType";
import BarkMessageLevel from "../../src/model/enumeration/BarkMessageLevel";
import BarkMessageSound from "../../src/model/enumeration/BarkMessageSound";
import BarkEncryptionError from "../../src/model/error/BarkEncryptionError";
import type BarkMessage from "../../src/model/request/BarkMessage";

const client = new BarkClient();

const mockAxios = new MockAdapter(axios);

afterEach(() => {
  mockAxios.reset();
});

describe("Health", () => {
  test("Server is healthy", async () => {
    mockAxios.onGet(BarkClientUrl.HEALTHZ).reply(200, "ok");

    await expect(client.health()).resolves.not.toThrowError();
  });

  test("Server is unhealthy", async () => {
    mockAxios.onGet(BarkClientUrl.HEALTHZ).timeout();

    await expect(client.health()).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe("Info", () => {
  test("Server respond info", async () => {
    mockAxios.onGet(BarkClientUrl.INFO).reply(200, {
      arch: "linux/arm64",
      build: "2023-04-10 08:13:23",
      commit: "dc8de8416c9c2c6c8cd6b95a85ff09c5653dfd11",
      devices: 1,
      version: "v2.1.5",
    });

    await expect(client.info()).resolves.toEqual({
      arch: "linux/arm64",
      build: new Date(Date.parse("2023-04-10 08:13:23")),
      commit: "dc8de8416c9c2c6c8cd6b95a85ff09c5653dfd11",
      devices: 1,
      version: "v2.1.5",
    });
  });

  test("Server don't respond info", async () => {
    mockAxios.onGet(BarkClientUrl.INFO).timeout();

    await expect(client.info()).rejects.toThrowErrorMatchingSnapshot();
  });
});

describe("Ping", () => {
  test("Server is running", async () => {
    mockAxios.onGet(BarkClientUrl.PING).reply(200, {
      code: 200,
      message: "pong",
      timestamp: Date.parse(new Date().toString()),
    });

    await expect(client.ping()).resolves.not.toThrowError();
  });

  test("Server is not running", async () => {
    mockAxios.onGet(BarkClientUrl.PING).timeout();

    await expect(client.info()).rejects.toThrowErrorMatchingSnapshot();
  });
});

const barkMessageCommonProperty = {
  badge: 1,
  body: "Test body",
  category: "Test category",
  copy: "Test copy",
  device_Key: "Device Key",
  group: "Bark SDK Unit Test",
  level: BarkMessageLevel.ACTIVE,
  sound: BarkMessageSound.ALARM,
  title: "Test title",
};

describe.each([
  {
    badge: barkMessageCommonProperty.badge,
    body: barkMessageCommonProperty.body,
    category: barkMessageCommonProperty.category,
    copy: barkMessageCommonProperty.copy,
    device_key: barkMessageCommonProperty.device_Key,
    isArchive: "1",
    group: barkMessageCommonProperty.group,
    level: barkMessageCommonProperty.level,
    sound: barkMessageCommonProperty.sound,
    title: barkMessageCommonProperty.title,
  },
  new BarkMessageBuilder()
    .archive()
    .badge(barkMessageCommonProperty.badge)
    .body(barkMessageCommonProperty.body)
    .category(barkMessageCommonProperty.category)
    .copy(barkMessageCommonProperty.copy)
    .deviceKey(barkMessageCommonProperty.device_Key)
    .group(barkMessageCommonProperty.group)
    .level(barkMessageCommonProperty.level)
    .sound(barkMessageCommonProperty.sound)
    .title(barkMessageCommonProperty.title)
    .build(),
])("Push $#", (barkMessage: BarkMessage) => {
  describe("Normal Push", () => {
    test("Device key is empty", async () => {
      delete barkMessage.device_key;

      mockAxios.onPost(BarkClientUrl.PUSH, barkMessage).reply(400, {
        code: 400,
        message: "device key is empty",
        timestamp: 0,
      });

      await expect(
        client.push(barkMessage),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    test("Device key is not registered", async () => {
      barkMessage.device_key = "I am not a device key";

      mockAxios.onPost(BarkClientUrl.PUSH, barkMessage).reply(400, {
        code: 400,
        message:
          "failed to get device token: failed to get [] devices token from database",
        timestamp: 0,
      });

      await expect(
        client.push(barkMessage),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    test("Request bind failed", async () => {
      mockAxios.onPost(BarkClientUrl.PUSH, barkMessage).reply(400, {
        code: 400,
        message: `request bind failed: invalid character '\\"' after object key:value pair`,
        timestamp: 0,
      });

      await expect(
        client.push(barkMessage),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    test("Push failed", async () => {
      mockAxios.onPost(BarkClientUrl.PUSH, barkMessage).reply(500, {
        code: 500,
        message: "push failed: ",
        timestamp: 0,
      });

      await expect(
        client.push(barkMessage),
      ).rejects.toThrowErrorMatchingSnapshot();
    });

    test("Succeed", async () => {
      mockAxios.onPost(BarkClientUrl.PUSH, barkMessage).reply(200, {
        code: 200,
        message: "success",
        timestamp: 0,
      });

      await expect(client.push(barkMessage)).resolves.not.toThrow();
    });
  });

  describe("Encrypted push", () => {
    delete barkMessage.device_key;

    const algorithm = BarkEncryptedPushAlgorithm.AES_128_CBC;
    const key = crypto.randomBytes(8).toString("hex");
    const iv = crypto.randomBytes(8).toString("hex");

    test.each([
      {
        algorithm: BarkEncryptedPushAlgorithm.AES_128_CBC,
        length: 16,
      },
      {
        algorithm: BarkEncryptedPushAlgorithm.AES_128_ECB,
        length: 16,
      },
      {
        algorithm: BarkEncryptedPushAlgorithm.AES_192_CBC,
        length: 24,
      },
      {
        algorithm: BarkEncryptedPushAlgorithm.AES_192_ECB,
        length: 24,
      },
      {
        algorithm: BarkEncryptedPushAlgorithm.AES_256_CBC,
        length: 32,
      },
      {
        algorithm: BarkEncryptedPushAlgorithm.AES_256_ECB,
        length: 32,
      },
    ])(
      "The length of key is not $length when algorithm is $algorithm",
      async ({ algorithm, length }) => {
        await expect(
          client.pushEncrypted(
            barkMessageCommonProperty.device_Key,
            barkMessage,
            algorithm,
            crypto.randomBytes(1).toString("hex"),
            iv,
          ),
        ).rejects.toThrowError(
          new BarkEncryptionError(
            BarkEncryptionErrorType.KEY_IS_NOT_CORRECT,
            `The length of key is not ${length}`,
          ),
        );
      },
    );

    test("The length of iv is not 16", async () => {
      await expect(
        client.pushEncrypted(
          barkMessageCommonProperty.device_Key,
          barkMessage,
          algorithm,
          key,
          crypto.randomBytes(1).toString("hex"),
        ),
      ).rejects.toThrowError(
        new BarkEncryptionError(
          BarkEncryptionErrorType.IV_IS_NOT_CORRECT,
          "The length of iv is not 16",
        ),
      );
    });
  });
});
