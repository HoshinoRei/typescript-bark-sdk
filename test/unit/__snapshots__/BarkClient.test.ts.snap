// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Health Server is unhealthy 1`] = `"Server has not response"`;

exports[`Info Server don't respond info 1`] = `"Server has not response"`;

exports[`Ping Server is not running 1`] = `"Server has not response"`;

exports[`Push 0 Normal Push Device key is empty 1`] = `"Device key is empty"`;

exports[`Push 0 Normal Push Device key is not registered 1`] = `"Failed to get device token: failed to get [] devices token from database"`;

exports[`Push 0 Normal Push Push failed 1`] = `"Push failed: "`;

exports[`Push 0 Normal Push Request bind failed 1`] = `"Request bind failed: invalid character '\\"' after object key:value pair"`;

exports[`Push 1 Normal Push Device key is empty 1`] = `"Device key is empty"`;

exports[`Push 1 Normal Push Device key is not registered 1`] = `"Failed to get device token: failed to get [] devices token from database"`;

exports[`Push 1 Normal Push Push failed 1`] = `"Push failed: "`;

exports[`Push 1 Normal Push Request bind failed 1`] = `"Request bind failed: invalid character '\\"' after object key:value pair"`;
